sensor/kubernetes/listener/resources/virtualmachine/store/virtual_machine_store.go

```diff
@@ -0,0 +1,220 @@
+}
+
+// Copy returns a copy of the VirtualMachineInfo
+func (v *VirtualMachineInfo) Copy() *VirtualMachineInfo {
+	if v == nil {
+		return nil
+	}
+	ret := &VirtualMachineInfo{
+		UID:       v.UID,
+		Name:      v.Name,
+		Namespace: v.Namespace,
+		Running:   v.Running,
+	}
+	if v.VSOCKCID != nil {
+		vsockCIDValue := *v.VSOCKCID
+		ret.VSOCKCID = &vsockCIDValue
+	}
+	return ret
+}
+
+// VirtualMachineStore stores the information about the VirtualMachines in the cluster
+type VirtualMachineStore struct {
+	lock sync.RWMutex
+
+	cidToUID        map[uint32]string
+	uidToCID        map[string]uint32
+	namespaceToUID  map[string]set.StringSet
+	virtualMachines map[string]*VirtualMachineInfo
+}
+
+// NewVirtualMachineStore returns a new store
+func NewVirtualMachineStore() *VirtualMachineStore {
+	return &VirtualMachineStore{
+		virtualMachines: make(map[string]*VirtualMachineInfo),
+		namespaceToUID:  make(map[string]set.StringSet),
+		cidToUID:        make(map[uint32]string),
+		uidToCID:        make(map[string]uint32),
+	}
+}
+
+// AddOrUpdateVirtualMachine upserts a new VirtualMachine
+func (s *VirtualMachineStore) AddOrUpdateVirtualMachine(vm *VirtualMachineInfo) {
+	if vm == nil {
+		return
+	}
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	s.addOrUpdateVirtualMachineNoLock(vm)
+}
+
+// AddOrUpdateVirtualMachineInstance upserts a new VirtualMachineInstance
+func (s *VirtualMachineStore) AddOrUpdateVirtualMachineInstance(uid, namespace string, vsockCID *uint32, isRunning bool) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	s.addOrUpdateVirtualMachineInstanceNoLock(uid, namespace, vsockCID, isRunning)
+}
+
+// RemoveVirtualMachine removes a VirtualMachine
+func (s *VirtualMachineStore) RemoveVirtualMachine(uid string) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	s.removeVirtualMachineNoLock(uid)
+}
+
+// RemoveVirtualMachineInstance removes a VirtualMachineInstance
+func (s *VirtualMachineStore) RemoveVirtualMachineInstance(uid string) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	s.removeVirtualMachineInstanceNoLock(uid)
+}
+
+// Cleanup resets the store
+func (s *VirtualMachineStore) Cleanup() {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	s.virtualMachines = make(map[string]*VirtualMachineInfo)
+	s.namespaceToUID = make(map[string]set.StringSet)
+	s.cidToUID = make(map[uint32]string)
+	s.uidToCID = make(map[string]uint32)
+}
+
+// OnNamespaceDeleted removes the VirtualMachines in the given namespace
+func (s *VirtualMachineStore) OnNamespaceDeleted(namespace string) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	vmIDsByNamespace := s.namespaceToUID[namespace]
+	for vmID := range vmIDsByNamespace {
+		s.removeVirtualMachineNoLock(vmID)
+	}
+}
+
+// Get returns the VirtualMachineInfo associated with the given UID
+func (s *VirtualMachineStore) Get(uid string) *VirtualMachineInfo {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+	return s.virtualMachines[uid].Copy()
+}
+
+func (s *VirtualMachineStore) addOrUpdateVirtualMachineNoLock(vm *VirtualMachineInfo) {
```

Comment:
how about having a new struct for VM as it do not have vsockcid
```suggestion
func (s *VirtualMachineStore) addOrUpdateVirtualMachineNoLock(vm *VirtualMachine) {
```